docs/devguide/devguide.md

```diff
@@ -98,19 +100,24 @@ To properly track changes across several Modules, you need to follow the recipe.
 ### Recipe
 Split up the data flow for each direction, one call for reading only, one call for writing only.
 Keep in mind that the caller by definition is "left" in the module graph and the callee is "right". The callee is the end of a callback, but for this pattern this has nothing to do with the direction the data flows in, which results in the rules defined below.
+
+![bidir example](bidir.png)
+
 - set up a ```uint32_t version_DATA``` in each Module
-- create a ```CallerSlot``` for ```DATACallRead``` for modules reading the ```DATA```
-- (optional) create a ```CallerSlot``` for ```DATACallWrite``` for modules writing the ```DATA```
+- create a ```CallerSlot``` for ```DATACallRead``` for modules consuming the ```DATA```
+- (optional) create a ```CallerSlot``` for ```DATACallWrite``` for modules providing ```DATA```
+- create a ```CalleeSlot``` each for ```DATACallRead``` and ```DATACallWrite``` for modules holding the ```DATA```
 - Create a ```DATACallRead``` either instancing the ```core::GenericVersionedCall``` template, or making sure that the call can distinguish between the ```DATA``` version that was last **set** into the Call and that which was last **got** out of the Call
 - (optional) Create a ```DATACallWrite``` along the same lines
-- create a ```CalleeSlot``` for ```DATACallRead``` for modules consuming the ```DATA```
-- create a ```CallerSlot``` for ```DATACallWrite``` for modules supplying the ```DATA```
 
-A module with slots for both directions by convention should first execute the reading and then provide updates via writing.
+A module with slots for both directions by convention must, in this exact order:
+- execute the reading, overwriting its DATA with incoming update
+- apply logic that will modify the DATA (if parameters or user input cause this, for example)
+- provide DATA updates via writing.
 
 #### Usage: ```DATACallRead```
 - In the ```GetData``` callback, make sure you can *supply unchanged data very cheaply*.
```

Comment:
This is meant for the holder, right? If so, mention it. The same for the case below.